### DISCUSSION TITLE
Content-Length header mismatch on Vercel

### DISCUSSION THREAD

**Discussion by spenpo** [vercel/community#1767] [Help]
### Summary

I have a blog on my site that uses graphql, and I setup ApolloServer as a serverless function like [this](https://github.com/spope851/spenpo-next/blob/main/pages/api/graphql.ts). I'm wondering if it's possible to convert my blog features to SSG with this stack? There are a few clear challenges but even after navigating them locally Vercel is still having trouble deploying the app

### Additional information

I'm trying to use the [with-apollo](https://github.com/vercel/next.js/blob/canary/examples/with-apollo/lib/apolloClient.js) approach to access the ApolloClient  in `getStaticProps`. [here](https://github.com/spope851/spenpo-next/blob/feature/static-gen/pages/blog/%5Bid%5D.tsx)'s an example. Initially, the problem was that the ApolloServer is only running when the frontend is deployed, so it's not accessible during build time. I had configured ApolloClient to access ApolloServer like this:
```js
const url = process.env.VERCEL_URL

const httpLink = new HttpLink({
  uri: `http${url ? `s://${url}` : "://localhost:3000"}/api/graphql`
})

function createApolloClient() {
  return new ApolloClient({
    ssrMode: typeof window === "undefined",
    link: from([errorLink, httpLink]),
    cache: new InMemoryCache(),
  })
}
```
This works for SSR, but not SSG. However, when I point `uri` to my production environment I'm able to build the project locally and it successfully fetches the data for all the pages in my blog. Ignoring the limitations of this bandaid fix, the new problem is that it still fails to build on vercel. I get the following error from the deployment:
![image](https://user-images.githubusercontent.com/34104632/223737985-ba2e90d0-ce59-418c-9342-92fa263326e9.png)

content length is not being properly set in the request header of the proxy. I've tried disabling compression but that didn't fix it. can't find many other occurrences of this one online. has anyone faced the same issue?

### Example

https://vercel.com/spope851/spenpo-next/J5nsxfhcbjTGZXQnh1KBZgVgVnLT

**Comment by amyegan** [comment]
Hi @spope851. It looks like you're trying to deploy your backend API and frontend all at once, which results in the issues you have seen where the backend is not yet available at build time. 

The [with-apollo](https://github.com/vercel/next.js/blob/canary/examples/with-apollo/lib/apolloClient.js) uses a [different project for it's backend](https://github.com/vercel/next.js/tree/canary/examples/with-apollo#:~:text=This%20example%20relies%20on%20Prisma%20%2B%20Nexus%20for%20its%20GraphQL%20backend.). If you deploy the API as a separate project, that should solve the problem 🙂

**Reply by spenpo** [reply]
Thank you! @amyegan 
I am actually still getting this same error when I try to initialize ApolloClient with my production server's backend, which is available at build time. The project builds successfully on my machine with this configuration, just not on Vercel

**Reply by smaeda-ks** [reply]
@spope851 What's your local Node.js version? It seems you're building the project with Node.js 18 on Vercel. The error comes from `undici`, which was added to Node.js core package as its `fetch` implementation since Node.js 18.

In fact, the same error would happen on my local builds with Node.js 18, but not with 16. Also, if I add the `useGETForQueries` option to `HttpLink`, then the error goes away. I'd encourage you to submit a bug report on https://github.com/apollographql/apollo-client for further help.

**Reply by spenpo** [reply]
@smaeda-ks thank you very much! as suspected, I'm running 16 locally. the culprit was the user agent undici which was the only observable difference in requests from the build in my local environment when compared with requests from the build in vercel's environment, as observed in the logs of my production server. the deployment passed when I added `useGetForQueries: true`. Thanks again!
